### DISCUSSION TITLE
Add `useQuery().promise` in order to support `React.use()`

### DISCUSSION THREAD

**Discussion by KATT** [TanStack/query#7980] [Ideas]
It's be nice to have something to support the `React.use()`-hook nicely. 

There might be a previous discussion about this (couldn't find one) or a good reason why this doesn't exist.

## Why?

It'd make it easier to parallelize queries as far up the tree as possible when doing SPAs.

## How

```tsx
import { useQuery } from '@tanstack/react-query';
import React from 'react';

function MyComponent() {
    const query = useQuery({
        // ...
    });

    const result = React.use(query.promise); 
    //      ^? Awaited<typeof promise>, or potentially `UseSuspenseQueryResult<Awaited<typeof promise>>`

    // [...]
}
```

### Implementation details

- The returned `Promise` needs to be stable so it doesn't change with every render, otherwise `use()` will cause infinite re-renders.
- Same sort of promise would be good for `useInfiniteQuery()` too

### Alternative API 

```tsx
import { usePromise, useInfinitePromise } from '@tanstack/react-query';

function MyComponent() {
    const promise1 = usePromise({ /* .... */ }); 
    const promise2 = useInfinitePromise({ /* .... */ });

    // [...]
}
```

## Example usage

```tsx
import { usePromise } from '@tanstack/react-query';
import React from 'react';
import { Comment, Post } from './somewhere';

function PostComments(props: {
    comments: Promise<Comment[]>;
}) {
    const comments = React.use(props.comments);

    return (
        <div>
            {comments.map(comment => (
                <Comment comment={comment} key={comment.id} />
            ))}
        </div>
    );
}

function Post(props: {
    comments: Promise<Comment[]>;
    post: Promise<Post>;
}) {
    const post = React.use(props.post);


    return (
        <div>
            <h1>{post.title}</h1>
            <p>{post.body}</p>
            <h2>
                Comments
                <Suspense fallback={<div>Loading comments...</div>}>
                    <PostComments comments={props.comments} />
                </Suspense>
            </h2>
        </div>
    )
    
}

export default function App() {
    const comments = useQuery({
        // ...
    }).promise;
    const post = useQuery({
        // ...
    }).promise;
    return (
        <Suspense fallback={<div>Loading post...</div>}>
            <Post post={post} comments={comments} />
        </Suspense>
    );
}
```

### Workaround today

The way I work around this today is that I have a `useQuery()` at the top of the tree and `useSuspenseQuery()`s further down the tree, using the same arguments. This is a bit of a hack, but it works.

**Comment by TkDodo** [comment]
Yes, I would want that. No, I don't really know how to do it 😅 .

API wise, I think the promise should just unwrap data:

```
const query = useQuery({
  queryKey: ['example'],
  queryFn: Promise.resolve(5),
})

const data = React.use(query.promise); 
//    ^ number
```

This would work great for initial fetches, and can also be used inside conditions like if statements. I'm just wondering how this would work:
- with queries that are `enabled: false` ? Would they just get a never resolving promise, or a promise that errors?
- with refetches. We generally shouldn't suspend again, but likely immediately resolve with stale data, and then eventually resolve with new data.

If you have an idea how to approach this, feel free to try a PR ❤️

**Reply by KATT** [reply]
Thanks for such a quick response!

> API wise, I think the promise should just unwrap data [...]

I can see use-cases where you'd like show a loading indicator that the data is being fetched (and have access to `.fetchNextPage()` etc), so I think the fuller `xxResult`-type might be preferable.


> with queries that are `enabled: false` ? Would they just get a never resolving promise, or a promise that errors?

Good question. Both alternatives would have good arguments. 

Either way, I think that stale data should be returned immediately akin to how `useQuery()` works with `enabled: false`

I'm leaning towards a promise that never resolves, but I don't have a strong conviction. 

> If you have an idea how to approach this, feel free to try a PR ❤️

The promise caching is tricky to get right.

- I think any pending query would have to exist as a `Promise` in the `QueryCache` 
- Any stale data should use the previous `Promise` as well and not the "latest one" - it gotta be updated after it resolves so any pending refetch doesn't trigger the `Suspense`

**Reply by TkDodo** [reply]
> I can see use-cases where you'd like show a loading indicator that the data is being fetched (and have access to .fetchNextPage() etc), so I think the fuller xxResult-type might be preferable.

you would get those right on `useQuery`:

```
// data is T | undefined here
const { isLoading, fetchNextPage, promise, data } = useInfiniteQuery(...)

if (...) {
  // realData is T here
  const realData = use(promise)
}
```

> The promise caching is tricky to get right.

We already cache the promise on the `Query` class.

> Any stale data should use the previous Promise

Good point, we would need to add that

**Reply by KATT** [reply]
I started playing a bit in #7988 & feeling like this:

![i_have_no_idea_what_im_doing_meme_640_07](https://github.com/user-attachments/assets/e53c60fe-acc7-4c9e-a6ce-ce5847561632)

**Comment by juliusmarminge** [comment]
> It'd make it easier to parallelize queries as far up the tree as possible when doing SPAs.

Isn't this what `usePrefetchQuery` (https://tanstack.com/query/latest/docs/framework/react/reference/usePrefetchQuery#useprefetchquery) is for?

Doesn't solve the duplicate (_a useQuery() at the top of the tree and useSuspenseQuery()s further down the tree, using the same arguments_) part though

**Reply by KATT** [reply]
The ergonomics is better with `use()`. It's too easy to forget about prefetching.

The idea is that your page will handle all the `.useQuery()` as far up as possible and any component will receive either...

- ```tsx 
  { promise: Promise<RouterOutputs['post']['byId']> } 
  ````

or

- ```tsx
  { query: ReactQueryResult['post']['byId'] }
  ```

.. as props
